python/ray/dag/compiled_dag_node.py

```diff
@@ -496,6 +603,145 @@ def exec_operation(
+            self.set_stream_buffer(op, input_data)
+        except RayChannelError:
+            # Channel closed. Exit the loop.
+            exit = True
+        return exit
+
+    def _stream_compute(self, op: _DAGNodeOperation, exec_stream, class_handle) -> bool:
+        """
+        Retrieve the intermediate result from the READ operation and perform the
+        computation on the provided execution stream. Then, cache the new
+        intermediate result.
+
+        Args:
+            op: The compute operation.
+            exec_stream: The CUDA stream to execute the compute operation.
+            class_handle: An instance of the class to which the actor belongs. For
+                example, the type of `class_handle` is <class 'xxxx.Worker'> if the
+                actor belongs to the `class Worker` class.
+
+        Returns:
+            True if system error occurs and exit the loop; otherwise, False.
+        """
+        input_data = self.reset_stream_buffer(op)
+        method = getattr(class_handle, self.method_name)
+        try:
+            _process_return_vals(input_data, return_single_output=False)
+        except Exception as exc:
+            # Previous task raised an application-level exception.
+            # Propagate it and skip the actual task. We don't need to wrap the
+            # exception in a RayTaskError here because it has already been wrapped
+            # by the previous task.
+            self.set_stream_buffer(exc)
+            return False
+
+        channel_results = []
+        for entry in input_data:
+            if isinstance(entry, tuple):
+                channel_result, event = entry
+                if event:
+                    event.synchronize()
+            else:
+                channel_result = entry
+            channel_results.append(channel_result)
+
+        resolved_inputs = []
+        for task_input in self.task_inputs:
+            resolved_inputs.append(task_input.resolve(channel_results))
+
+        import cupy as cp
+
+        exec_event = cp.cuda.Event()
+        with exec_stream:
+            try:
+                output_val = method(*resolved_inputs, **self.resolved_kwargs)
+            except Exception as exc:
+                output_val = _wrap_exception(exc)
+            exec_event.record()
+
+        self.set_stream_buffer(op, (output_val, exec_event))
+        return False
+
+    def _stream_write(self, op: _DAGNodeOperation) -> bool:
+        """
+        Retrieve the intermediate result from the COMPUTE operation and write to its
+        downstream DAG nodes. The caller must ensure that the last operation executed
+        is COMPUTE so that the function retrieves the correct intermediate result.
+
+        Returns:
+            True if system error occurs and exit the loop; otherwise, False.
+        """
+        output_val, exec_event = self.reset_stream_buffer(op)
+        exit = False
+        exec_event.synchronize()
```

Comment:
maybe we should not block cpu here.

Reply:
this way, we can overlap shm write and compute (if it runs in kernel)

Reply:
I think we need to sync to make sure execution finishes, otherwise the value may be incorrect?

Reply:
I think you can synchronize on compute stream, and then torch and cuda should handle the gpu synchronization

Reply:
It's better to sync on the event since there may be additional operations on the compute stream? Wouldn't waiting on the compute stream require unnecessarily longer waiting time?

Could you elaborate more on "then torch and cuda should handle the gpu synchronization"? Not sure what it means.

Reply:
hmm actually my bad. I think if event is created with blocking=False (which is the default https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Event.html), this only blocks the relevant device, not cpu. So I think the current code is fine.